### PR TITLE
Refactor jerk-limited trajectory generator

### PR DESCRIPTION
**Pull request by bresch**
### Context
Currently, algorithm generates the trajectory by performing a numerical integration of the current jerk. This is relatively fast to compute but has a few drawbacks:
- requires small dt to be accurate: slowing down the algorithm creates nasty numerical issues
- limits the maximum jerk: if the desired jerk is too large, the time where the jerk is active can be smaller than dt, this can freeze the trajectory
- sensitive to dt jitter: if the integration uses a different dt than the one used to predict the timings, slight imperfections will occur and destabilize the algorithm
- fails to compute accurate results around zero: the algorithm struggles to reach exactly zero speed

All those "corner cases" are handled to make the algorithm viable but the benefit of this light weight numerical integration with all those fixes isn't really relevant anymore.

### New implementation
Evaluating the piecewise polynomials for the current time instead of the numerical integration gives the exact result regardless time jitter, jerk amplitude and converges properly to small values without the need of tricks.
- It is now possible to set large jerk and accelerations without any numerical issues
- The overall implementation is simpler, no corner case handling
- The result is exact

### Comparison
**Current algorithm**
vel_traj_graph

**New proposed algorithm**
New_traj_gen

The unit test has also been updated to verify the time synchronization between two trajectories, the exactness of the predicted time and a few trivial cases.

TODO:
- [x] test auto mode
- [x] reintroduce time stretching in auto mode
- [x] test altitude mode
- [x] test manual mode

### CODE REVIEW

**Comment by bresch** [comment]
Tested in SITL, example of log with mission, RTL, manual position and altitude modes: https://logs.px4.io/plot_app?log=d27b2e59-a6e6-48ae-bfc2-c12b7559ec8b

I don't see regression, only improvements (no more velocity overshoot)

**Comment by bresch** [comment]
@PX4/testflights Can you fly this PR please?
- Altitude mode:
   - fly around, change altitude
   - up and down motion, wait sometimes a few seconds between motions
- Position control: 
   - move around
   - change altitude
   - start and stop
   - hold position for 2-3 minutes
- Auto:
   - Takeoff
   - Mission
   - Land
   - RTL
   - goto

And everything else you have on your list.

Thanks in advance!

**Comment by dannyfpv** [comment]
Tested on Pixhawk 4 v5
Modes Tested

Position Mode and Altitude Mode Stabilized Mode: on position  mode the copter go back to home point and loiter
Mission Plan Mode (Automated): no  issue
RTL: Good.


Notes: go back to the home point and loiter

altitude  and  position  mode:
https://review.px4.io/plot_app?log=dce64c36-4944-4187-8627-5894ce62f3d2

mission mode:
https://review.px4.io/plot_app?log=895a0c5b-3266-4cf2-8230-5a5e53540c23


Tested on Pixhawk 4 mini v5
Modes Tested

Altitude mode:
https://review.px4.io/plot_app?log=4f8cd02c-b494-4099-bc4a-759d9c1ee70e

Position control:
https://review.px4.io/plot_app?log=159e309b-56e9-4204-aa2e-004e304c4c84

Auto:
https://review.px4.io/plot_app?log=62b5589a-91e4-46e5-b1ee-1255fa533cb2


Tested on Pixracer  v4
Modes Tested


position  mode:
https://review.px4.io/plot_app?log=60199843-e026-4887-aa87-f88d3c40be1b

**Comment by bresch** [comment]
Rebased on master.

@jorge789 Thanks, after inspecting the log of the V5 I don't see any issue.
> I just noticed the commit is the same as yesterday for the Pixracer. I downloaded the firmware really early this morning. I will test again tomorrow again just to confirm.

Okay, good.

**Comment by dannyfpv** [comment]
Tested on PixRacer V4:
Mode tested:
Position Mode

Note: no issues on position mode

log:
https://review.px4.io/plot_app?log=f271867c-fabf-4c9d-a19f-942b2dc09d66


Tested on Pixhawk4 V5:
Mode tested:
Position Mode

Note: no issues on position mode

log:
https://review.px4.io/plot_app?log=ded6f199-bec8-4adb-98c8-41c0dff9ef20

**Comment by bresch** [comment]
Rebased on master

File: src/lib/FlightTasks/tasks/Utility/VelocitySmoothing.cpp [hunk 1]
```diff
@@ -171,112 +154,132 @@ float VelocitySmoothing::computeT2(float T123, float T1, float T3)
...
+		t1 = _T1;
+		t2 = _T2;
+		t3 = _T3;
+		t4 = t - _T1 - _T2 - _T3;
 	}
 
-	/* Integrate the trajectory */
-	float accel_new, vel_new, pos_new;
-	integrateT(dt * integration_scale_factor, _jerk, _accel, _vel, _pos, accel_new, vel_new, pos_new);
+	if (t > 0.f) {
```
Comment (MaEtUgR) [inline src/lib/FlightTasks/tasks/Utility/VelocitySmoothing.cpp]: You could move these ifs as `(t_remain > 0)` in between the steps of my suggestion for the `t` split up and return early after the first instance the condition was true.

File: src/lib/FlightTasks/tasks/Utility/VelocitySmoothing.cpp [hunk 2]
```diff
@@ -171,112 +154,132 @@ float VelocitySmoothing::computeT2(float T123, float T1, float T3)
...
+	return traj;
+}
 
-	} else if (_T3 > FLT_EPSILON) {
-		_jerk = -_max_jerk_T1;
+void VelocitySmoothing::updateTraj(float dt, float time_stretch)
+{
+	_local_time += dt * time_stretch;
+	const float t = _local_time;
+	float t1 = 0.f;
```
Comment (MaEtUgR) [inline src/lib/FlightTasks/tasks/Utility/VelocitySmoothing.cpp]: With the `t` spiltup I suggested you could initialize these when you know the value and only if they are needed.

**Comment by bresch** [comment]
@MaEtUgR Thanks a lot for the review and suggestions. I've made some improvements, could you review again and approve if it looks good to you please?
